src/config/base.cpp

```diff
@@ -94,59 +190,159 @@ ConfigBase::ConfigBase(std::string_view dump) {
 }
 
+ConfigBase::~ConfigBase() {
+    sodium_free(_keys);
+}
+
+int ConfigBase::key_count() const {
+    return _keys_size;
+}
+
+bool ConfigBase::has_key(ustring_view key) const {
+    if (key.size() != 32)
+        throw std::invalid_argument{"invalid key given to has_key(): not 32-bytes"};
+
+    auto* keyptr = key.data();
+    for (size_t i = 0; i < _keys_size; i++)
+        if (sodium_memcmp(keyptr, _keys[i].data(), KEY_SIZE) == 0)
+            return true;
+    return false;
+}
+
+std::vector<ustring_view> ConfigBase::get_keys() const {
+    std::vector<ustring_view> ret;
+    ret.reserve(_keys_size);
+    for (size_t i = 0; i < _keys_size; i++)
+        ret.emplace_back(_keys[i].data(), _keys[i].size());
+    return ret;
+}
+
+void ConfigBase::add_key(ustring_view key, bool high_priority) {
+    static_assert(
+            sizeof(Key) == KEY_SIZE, "std::array appears to have some overhead which seems bad");
+
+    if (key.size() != KEY_SIZE)
+        throw std::invalid_argument{"add_key failed: key size must be 32 bytes"};
+
+    if (_keys_size > 0 && sodium_memcmp(_keys[0].data(), key.data(), KEY_SIZE) == 0)
+        return;
+    else if (!high_priority && has_key(key))
+        return;
+
+    if (_keys_capacity == 0) {
+        // There's not a lot of point in starting this off really small: sodium is likely going to
+        // use at least a page size anyway.
+        _keys_capacity = 16;
+        _keys = static_cast<Key*>(sodium_allocarray(_keys_capacity, KEY_SIZE));
+    }
+
+    if (_keys_size >= _keys_capacity) {
+        _keys_capacity *= 2;
+        auto new_keys = static_cast<Key*>(sodium_allocarray(_keys_capacity, 32));
+        if (high_priority) {
+            std::memcpy(new_keys[0].data(), key.data(), KEY_SIZE);
+            std::memcpy(&new_keys[1], _keys, _keys_size * KEY_SIZE);
+        } else {
+            std::memcpy(&new_keys[0], _keys, _keys_size * KEY_SIZE);
+            std::memcpy(new_keys[_keys_size].data(), key.data(), KEY_SIZE);
+        }
+        sodium_free(_keys);
+        _keys = new_keys;
+    } else if (high_priority) {
+        // shift everything up so we can insert at beginning
+        std::memmove(&_keys[1], &_keys[0], _keys_size * KEY_SIZE);
+        std::memcpy(_keys[0].data(), key.data(), KEY_SIZE);
+    } else {
+        // add at the end
+        std::memcpy(_keys[_keys_size].data(), key.data(), KEY_SIZE);
+    }
+    _keys_size++;
+
+    // *Slightly* suboptimal in that we might change buffers above even when we didn't need to, but
+    // not worth worrying about optimizing.
+    if (high_priority)
+        remove_key(key, 1);
+}
+
+int ConfigBase::clear_keys() {
+    int ret = _keys_size;
+    _keys_size = 0;
+    return ret;
+}
+
+bool ConfigBase::remove_key(ustring_view key, size_t from) {
+    bool removed = false;
+
+    for (size_t i = from; i < _keys_size; i++) {
+        if (sodium_memcmp(_keys[0].data(), _keys[i].data(), KEY_SIZE) == 0) {
```

Comment:
should this be `if (sodium_memcmp(key.data(), _keys[i].data(), KEY_SIZE) == 0) {` instead ?

Reply:
Yes indeed it should, good catch!